### PR TITLE
Synthetic data generates on demand rather than on schedule

### PR DESCRIPTION
**Pull request by yaroslavyaroslav**
### TLDR

This PR planned to provide on demand synthetic data calculation, where on demands stands for in addition to request received.

### Implementation details

This PR consists from two parts:
- The main one, like the logic of heavy computational tasks caching and the modified logic of the scheduling from schedule loop to expanding data of any given task received. 
- And the work around one, that is required to make this behavior feasible in isolated environment (with nodes, miners and all the stuff disabled).

#### Main part of PR

Main part consists of the following changes:

- `refresh_sythetic_data.py` has been (well it actually wasn't completely but supposed to be) rewritten to store nothing but a cache warmup function.
- `schedule_synthetic_queries.py` is the main place where the most changes have been made
    - `_fetch_tasks_in_chunks` function reads a chunk of new organic tasks to proceed, if any.
    - `_enhance_task` function enhances the initial raw task with the unique data. Currently implemented just for `chat` and `text-to-image` tasks, because of it's nothing more but a switch by task kind it can be expanded effortlessly.
    - `_push_to_query_node` function pushes enhanced tasks to a `query_node`
    - `warmup_function` function implements logic of a service warmup in sake of code intents clarification
    - `schedule_synthetics_until_done` function has been rewritten to reflect the new approach — enhance any given tasks amount with synthetic data
- `execution_cycle.py` — `warmup_function` call added on a service start.

#### Work around part of PR

- `cli.py` cli tool that allows developer to manually send arbitrary amount of tasks of any kind to the `control_node` to check the whole `-> contol_node -> query_node` pipeline.
- `redis_constants.py` — `CONTROL_NODE_QUEUE_KEY` and `CONTROL_TASK_CHUNK_SIZE` constants have been added to be used in `cli.py` and in the queue that is modified `schedule_synthetics_until_done` function reads raw tasks from.

Closes: #55

### CODE REVIEW

File: validator/control_node/src/cycle/execute_cycle.py [hunk 1]
```diff
@@ -43,25 +44,25 @@ async def _post_vali_stats(config: Config):
     )
 
 
-async def get_nodes_and_contenders(config: Config) -> list[Contender] | None:
+async def get_nodes_and_contenders(config: Config) -> List[Contender] | None:
```
Comment (namoray) [inline validator/control_node/src/cycle/execute_cycle.py:47]: Why refactor typehints? Prefer the lowercase

File: validator/control_node/src/cycle/execute_cycle.py [hunk 2]
```diff
@@ -43,25 +44,25 @@ async def _post_vali_stats(config: Config):
...
     logger.info("Starting cycle...")
     if config.refresh_nodes:
-        logger.info("First refreshing metagraph and storing the nodes")
```
Comment (namoray) [inline validator/control_node/src/cycle/execute_cycle.py:49]: Why is all this refactored?

File: validator/control_node/src/cycle/execute_cycle.py [hunk 3]
```diff
@@ -70,29 +71,25 @@ async def main(config: Config) -> None:
     time_to_sleep_if_no_contenders = 20
     contenders = await get_nodes_and_contenders(config)
 
-    # date_to_delete = datetime(2024, 10, 13, 30)
```
Comment (namoray) [inline validator/control_node/src/cycle/execute_cycle.py:73]: Why is this refactored?

File: validator/control_node/src/cycle/execute_cycle.py [hunk 4]
```diff
@@ -70,29 +71,25 @@ async def main(config: Config) -> None:
...
-    # async with await config.psql_db.connection() as connection:
-    #     await delete_task_data_older_than_date(connection, date_to_delete)
-    #     await delete_contender_history_older_than(connection, date_to_delete)
-    #     await delete_reward_data_older_than(connection, date_to_delete)
+    await warmup_function(config=config)
```
Comment (namoray) [inline validator/control_node/src/cycle/execute_cycle.py:74]: 'warmup_function' is a bad name -> what is it doing?

File: validator/control_node/src/cycle/schedule_synthetic_queries.py [hunk 5]
```diff
@@ -17,19 +17,32 @@
...
+@dataclass(order=True)
 class TaskScheduleInfo:
+    next_schedule_time: float
     task: str
     total_requests: int
     interval: float
-    next_schedule_time: float
     remaining_requests: int
 
-    def __lt__(self, other: "TaskScheduleInfo"):
```
Comment (namoray) [inline validator/control_node/src/cycle/schedule_synthetic_queries.py:31]: Why is this refactoreD?

File: validator/control_node/src/cycle/schedule_synthetic_queries.py [hunk 6]
```diff
@@ -107,94 +126,96 @@ async def _clear_old_synthetic_queries(redis_db: Redis):
...
+    redis_client = config.redis_db
+
+    while True:
+        tasks = await _fetch_tasks_in_chunks(redis_client, rcst.CONTROL_TASK_CHUNK_SIZE)
+
+        if not tasks:
+            logger.info("No more tasks to fetch, exiting the scheduling loop.")
             break
 
-        if i % 100 == 0:
```
Comment (namoray) [inline validator/control_node/src/cycle/schedule_synthetic_queries.py:172]: Why has all this logic gone? We've completely changed how we schedule synthetic queries now right? All this logic must remain

**Review by namoray — CHANGES_REQUESTED** [review]
it's really hard to review this PR because a massive amount of code has been refactored which (a) did not need to be, and (b) is not relevant to the changes we're tryng to make here

I'm also not sure what the PR description means, or how this solution solves the problem. I believe it changes functionality of the whole system too much
